File: php/init/access-jvm.php
<?php
// 通过 use 导入java类, 通过"m8test_java\全类名"的方式
use m8test_java\java\lang\StringBuilder;
use m8test_java\android\view\Gravity;
use m8test_java\java\lang\System;

/** @var com.m8test.script.core.api.extension\Extensions $extensions */
/** @var com.m8test.script.core.api.file\Files $files */

$logger->info("access-jvm.php started");

$dynamicJava = $files->buildFile(function ($builder) {
    $builder->setRelativePath("", "res/com/example/script/DynamicJvmAccess.java");
});
$extensions->loadJavaFile($dynamicJava);
importJavaClass("com.example.script.DynamicJvmAccess", "DynamicJvmAccess");
// 通过 new 创建java对象
$sb = new StringBuilder("M8Test");
/** @var m8test_java\com\m8test\script\core\api\logger\Logger $logger */
// 调用java对象方法使用 ->
$sb->append("Php");
$logger->info("StringBuilder: " . $sb->toString());
// 调用java静态方法使用 ::
$logger->info("currentTimeMillis: " . System::currentTimeMillis());
// 调用java静态属性使用 :: 并且添加前缀 $_MJ_ , 编译后会变成 Gravity::CENTER
$logger->info("Gravity.CENTER: " . Gravity::$_MJ_CENTER);
// 如果 :: 不能调用静态属性, 那么需要通过反射调用
/** @var m8test_java\com\m8test\script\core\api\reflect\Reflectors $reflectors */
$v = $reflectors->byName("com.example.script.DynamicJvmAccess")->getStaticField(function ($f) {
    $f->setName("KIND");
});
$logger->info("dynamic.KIND: " . $v);
$logger->info("dynamic.add: " . DynamicJvmAccess::add(2, 3));
$logger->info("dynamic.byte: " . DynamicJvmAccess::echoByte(8));
$logger->info("dynamic.short: " . DynamicJvmAccess::echoShort(16));
$logger->info("dynamic.int: " . DynamicJvmAccess::echoInt(32));
$logger->info("dynamic.long: " . DynamicJvmAccess::echoLong(64));
$logger->info("dynamic.float: " . DynamicJvmAccess::echoFloat(1.25));
$logger->info("dynamic.double: " . DynamicJvmAccess::echoDouble(2.5));
$logger->info("dynamic.char: " . DynamicJvmAccess::echoChar("Z"));
$logger->info("dynamic.boolean: " . DynamicJvmAccess::echoBoolean(true));
$logger->info("dynamic.byteBoxed: " . DynamicJvmAccess::echoByteBoxed(9));
$logger->info("dynamic.intBoxed: " . DynamicJvmAccess::echoIntBoxed(33));
$logger->info("dynamic.booleanBoxed: " . DynamicJvmAccess::echoBooleanBoxed(false));
$logger->info("dynamic.negate: " . DynamicJvmAccess::negate(true));
$logger->info("dynamic.scale: " . DynamicJvmAccess::scale(1.5, 2.0));
$logger->info("dynamic.join: " . DynamicJvmAccess::join("甲", "乙"));
$logger->info("dynamic.boxed: " . DynamicJvmAccess::boxed(7));
$logger->info("dynamic.identity: " . DynamicJvmAccess::identity("对象"));
$logger->info("dynamic.nullable: " . DynamicJvmAccess::nullable(null));
$logger->info("dynamic.date: " . DynamicJvmAccess::date());
$logger->info("dynamic.calendar: " . DynamicJvmAccess::calendar());
$logger->info("dynamic.bigInteger: " . DynamicJvmAccess::bigInteger());
$logger->info("dynamic.bigDecimal: " . DynamicJvmAccess::bigDecimal());
$logger->info("dynamic.byteArrayLength: " . DynamicJvmAccess::byteArrayLength($arrays->byteArrayOf(1, 2, 3)));
$stringClass = $classes->getJavaClassByName("java.lang.String");
$logger->info("dynamic.className: " . DynamicJvmAccess::className($stringClass));
$logger->info("dynamic.objectArrayLength: " . DynamicJvmAccess::objectArrayLength(
    $arrays->arrayOf($stringClass, "a", "b")
));
$logger->info("dynamic.sumInts: " . DynamicJvmAccess::sumInts($arrays->intArrayOf(1, 2, 3)));
$logger->info("dynamic.sumLongs: " . DynamicJvmAccess::sumLongs($arrays->longArrayOf(4, 5)));
$logger->info("dynamic.listSize: " . DynamicJvmAccess::listSize($iterables->listOf("x", "y")));
$logger->info("dynamic.setSize: " . DynamicJvmAccess::setSize($iterables->setOf("x", "y")));
$logger->info("dynamic.mapValue: " . DynamicJvmAccess::mapValue($maps->mapOf($maps->pairOf("answer", 42)), "answer"));
$logger->info("dynamic.varargs: " . DynamicJvmAccess::varargs("left", "right"));
$logger->info("dynamic.intArrayRoundtrip: " . DynamicJvmAccess::sumInts(DynamicJvmAccess::ints(4, 5)));
$logger->info("dynamic.ints: " . DynamicJvmAccess::ints(4, 5));
$logger->info("dynamic.strings: " . DynamicJvmAccess::strings("a", "b"));
$logger->info("dynamic.list: " . DynamicJvmAccess::list("x", "y"));
$logger->info("dynamic.map: " . DynamicJvmAccess::map("answer", 42));
$logger->info("dynamic.enum: " . DynamicJvmAccess::enumValue());
$logger->info("dynamic.sam: " . DynamicJvmAccess::callSam(function ($value) {
    return $value * 2;
}, 6));

File: python/init/AccessJvm.py
# -*- coding: utf-8 -*-
import os

from m8test_java.com.m8test.script.GlobalVariables import _logger, _files, _extensions, _java
from m8test_java.com.m8test.python.PythonJepBridge import PythonJepBridge

_logger.info("PYTHON_INIT_ACCESS_JVM_BEGIN")

# 动态加载随脚本项目推送的 Java 源文件。Java 文件必须由当前脚本显式加载，
# 这样可以验证外部 Java 源文件编译、Dex 转换和 ClassLoader 更新链路。
java_file = _files.buildFile(
    lambda builder: builder.setPath(
        os.path.abspath(os.path.join(
            os.path.dirname(__file__), "..", "res",
            "com", "example", "script", "PythonJvmAccess.java"
        ))
    )
)
PythonJepBridge().loadJavaFile(_extensions, java_file)
_logger.info("PYTHON_INIT_ACCESS_JVM_JAVA_FILE_LOADED")

PythonJvmAccess = _java.loadClass("com.example.script.PythonJvmAccess")
_logger.info(PythonJvmAccess.message())
_logger.info("PYTHON_INIT_ACCESS_JVM_STATIC_METHOD_DONE")
python_jvm_access = PythonJvmAccess()
_logger.info(python_jvm_access.instanceMessage())
_logger.info("PYTHON_INIT_ACCESS_JVM_INSTANCE_METHOD_DONE")
_logger.info(PythonJvmAccess.STATIC_FIELD)
_logger.info(python_jvm_access.instanceField)
_logger.info(str(PythonJvmAccess.sumInt(2, 3)))
_logger.info(str(PythonJvmAccess.sumLong(4, 5)))
_logger.info(str(PythonJvmAccess.sumNumbers(1, 2, 3.5, 4.25)))
_logger.info(PythonJvmAccess.describe("python", True, "!"))
_logger.info(str(PythonJvmAccess.arrayLength([1, 2, 3])))
_logger.info(PythonJvmAccess.joinStrings(["one", "two", "three"]))
_logger.info(PythonJvmAccess.listSummary(["first", "last"]))
_logger.info(PythonJvmAccess.callback(lambda value, count: value + ":" + str(count)))
_logger.info(python_jvm_access.instanceValues("instance", 9, False))
_logger.info(python_jvm_access.instanceArray([4, 5, 6]))

# 创建 java 对象
from m8test_java.java.lang.StringBuilder import StringBuilder

sb = StringBuilder("M8Test")
# 通过 . 调用 java 对象方法
sb.append("Python")
_logger.info(sb.toString())
# 通过 . 调用java静态方法
from m8test_java.java.lang.System import System

_logger.info(str(System.currentTimeMillis()))
# 可以通过`.`访问静态属性
from m8test_java.android.view.Gravity import Gravity

_logger.info(str(Gravity.CENTER))
# 如果不能通过`.`访问静态属性, 那么可以反射获取
from m8test_java.com.m8test.script.GlobalVariables import _reflectors
from m8test_java.android.widget.FrameLayout import FrameLayout

wrap_content = PythonJepBridge().getStaticField(
    _reflectors, FrameLayout.LayoutParams, "WRAP_CONTENT"
)
_logger.info(str(wrap_content))
# 访问java内部类可以使用 . , 必须使用'外部类.内部类'的方式调用
layout_params = FrameLayout.LayoutParams(wrap_content, wrap_content)
_logger.info(str(layout_params))
_logger.info("PYTHON_INIT_ACCESS_JVM_DONE")
